sftp/sftp.go

```diff
@@ -101,6 +94,81 @@ func (fm *fileManagerImpl) Delete(remoteFilePath string) error {
 	if err != nil {
 		return fmt.Errorf("cannot delete file: %w", err)
 	}
+	return nil
+}
 
+func (fm *fileManagerImpl) Reset() error {
+	newFm, err := NewFileManager(fm.config)
+	if err != nil {
+		return err
+	}
+	fm.client = newFm.client
 	return nil
 }
+
+func NewFileManager(config *SSHConfig) (*fileManagerImpl, error) {
+	sshClient, err := newSSHClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("creating SSH client: %w", err)
+	}
+	sftpClient, err := newSFTPClient(sshClient)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create SFTP client: %w", err)
+	}
+	return &fileManagerImpl{client: sftpClient, config: config}, nil
+}
+
+type retryableFileManagerImpl struct {
+	fileManager FileManager
+}
+
+func (r *retryableFileManagerImpl) Upload(localFilePath, remoteFilePath string) error {
+	fileOperation := func() error {
+		return r.fileManager.Upload(localFilePath, remoteFilePath)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Download(remoteFilePath, localDir string) error {
+	fileOperation := func() error {
+		return r.fileManager.Download(remoteFilePath, localDir)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Delete(remoteFilePath string) error {
+	fileOperation := func() error {
+		return r.fileManager.Delete(remoteFilePath)
+	}
+	return r.retryOnConnectionLost(fileOperation)
+}
+
+func (r *retryableFileManagerImpl) Reset() error {
+	return r.fileManager.Reset()
+}
+
+func (r *retryableFileManagerImpl) retryOnConnectionLost(fileOperation func() error) error {
```

Comment:
Here we are only retrying once, with no delay. Assuming a connection issue we might need a few more retries spaced out in time to handle it better.

I would suggest using backoff retry, like we do [here](https://github.com/rudderlabs/rudder-server/blob/1aadeb0ca355915bd7559c986f5df710bdc2bdfc/warehouse/integrations/clickhouse/clickhouse.go#L548)

Reply:
if that is required we can handle in client side
cc- @koladilip

Reply:
@lvrach this is not intend for general retry but retry when connection is closed due to idle time